### DISCUSSION TITLE
Validator throwing error when using code from CodeSystem where .content = 'not-present', should be warning?

### DISCUSSION THREAD

**Discussion by jd-nictiz** [FirelyTeam/firely-net-sdk#2229] [Ideas]
Hi!

First of all, apologies for lacking any specific background knowledge of the .NET SDK. I am a mere Simplifier end-user, raising this discussion from a Firely Service Desk ticket (SD-1132).

At Nictiz, the functional specification we base our profiles on (the zibs) prescribes the use of 'external' code systems, which we can't import in FHIR and which we don't want to, because we feel the party that maintains it should. However, to make the situation work, we create `CodeSystem`s with `.content` set to _not-present_. These `CodeSystem`s do therefore not contain any `.concept`s. For example, see https://simplifier.net/packages/nictiz.fhir.nl.r4.zib2020/0.3.0-beta1/files/593597/~xml 

As a consequence, the `ValueSet`s refering to these `CodeSystem`s include "the whole of the `CodeSystem`", also without specifying any `.concept`s. For example: https://simplifier.net/packages/nictiz.fhir.nl.r4.zib2020/0.3.0-beta1/files/593584/~xml

When validating a resource using a code from this `ValueSet`/`CodeSystem`, we get the following error:
`CodeInvalid : Code '71587' from system 'urn:oid:2.16.840.1.113883.2.4.4.1' does not exist in valueset 'http://decor.nictiz.nl/fhir/ValueSet/2.16.840.1.113883.2.4.3.11.60.121.11.12--20200901000000'`

We feel that this should be a warning instead of an error, with the message stating something in the spirit of "All codes from system 'XXX' are allowed, but this specific code 'XXX' could not be checked.". Because, in this case, the code '71587' just isn't invalid, it's just that you can't check it specifically, which is a whole different problem.

We understand from the Firely ticket that you state that this is a deliberate choice, stating it would be dangerous to report success (kind of paraphrasing here). In our opinion, throwing a warning does not equal success. Furthermore, doesn’t this statement make the `CodeSystem.content` element worthless for any other code than complete (https://hl7.org/fhir/r4/valueset-codesystem-content-mode.html)? From a pure FHIR perspective I can understand the viewpoint, but this just isn't the world we are living in - not every code system out there is captured in a correct `CodeSystem` resource (yet, hopefully).

Hope to get some discussion on this! Thanks

Jorn Duwel
Nictiz

**Comment by ewoutkramer** [comment]
Hi Jorn - let's see how we can improve this situation.  There are two major parts at work here:

* The validator - it gets the result from the terminology service, which is basically either "true" or "false".  And there is a result where the result is true, but the terminology service adds additional warning information. The validator will convert "false" to an error and true to either success or warning, depending on whether the terminology service returns additional information.  Important to realize is that the validator will *never* return errors for terminology unless the binding is of strength "required". If a binding is not required, the result is always success. Just to verify: is the binding we are talking about indeed required?
* The terminology service - it can basically only return true/false + an additional message.  It could/should take this `valueset-codesystem-content-mode` in consideration when validating - that is a good suggestion.  I think the Simplifier validator uses the basic  .NET SDK terminology service, which does *not* do this yet.  @mharthoorn - is this true for Simplifier?  @marcovisserFurore - we might need to take this extension into account in our SDK terminology service.

**Reply by jd-nictiz** [reply]
Hi Ewout. The binding we are talking about ineed is required. Thanks for the extra information, following the discussion.

**Reply by ewoutkramer** [reply]
Thanks for the confirmation - this means we should look at improving the built-in terminology service in Simplifier to 1) look for the extension you mentioned and 2) in general try to determine how complete the expansion is, before concluding that a code is not part of a coding system. This is the only way the validator can draw the right conclusions when dealing with required bindings.  @marcovisserFurore - assuming Simplifier is using the built-in SDK terminology service, shall we add this as an enhancement to our backlog?

**Comment by helmavdl** [comment]
Following this discussion I'd argue that it should neither be an error nor a warning. 
- In reality it will not always be possible to validate against a terminology server during the validation against a profile. Depending on the software architecture or simply because of business rules the validation of the codes happens much later in the process.
- Also, the profiles made by Nictiz are intended to be reused as much as possible to improve interoperability. That also means the valueSets and codeSystems used in the profiles. Some of these codeSystems are not publicly available due to licensing. Therefore the valueSets cannot be resolved.
- I am building examples to accompany our specifications and currently they do not validate due to these errors. This makes the examples useless as documentation. And examples are usually the best studied part of specifications.
- Based on past experience I also don't tolerate warnings as they can hide errors. Therefore, for me errors and warnings are equally bad.

I've tried a workaround: as a test I added a simple codeSystem to my package that defined the missing codes and CodeSystems. However, the error, nor the warnings went away.

My suggestion would be to validate the coding.system and throw an _informational_ notification when the coding.code cannot be validated.

**Reply by ewoutkramer** [reply]
I am pretty sure there will be users who would totally disagree that we should swallow and hide warnings about missing valuesystems!  

Would "informational" solve anything?  Depending on the error level it means you won't see them at all (I guess that's what you are aiming for, but see my statement above) or they are visible and will clog up your error reports.

**Comment by ewoutkramer** [comment]
We've improved the situation somewhat. There is now a `MultiTerminologyService` (https://docs.fire.ly/projects/Firely-NET-SDK/en/latest/validation/terminology-service.html#multiterminologyservice) - you can configure this service to route the terminology validation calls to a specific ITerminologyService, depending on the valueset.  So, to make this work, you can make a simple implementation of `ITerminologyService` that always returns success, and then configure the `MultiTerminologyService` to use this "always success" service for a specific set of valuesets. I think will cover your usecase!
